plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,565 @@
+                .build();
+        s3Client.deleteObjects(request);
+    }
+
+    /**
+     * Helper function used to work around the fact that if you use an S3 bucket with an '_' that java.net.URI
+     * behaves differently and sets the host value to null whereas S3 buckets without '_' have a properly
+     * set host field. '_' is only allowed in S3 bucket names in us-east-1.
+     *
+     * @param uri The URI from which to extract a host value.
+     * @return The host value where uri.getAuthority() is used when uri.getHost() returns null as long as no UserInfo is present.
+     * @throws IllegalArgumentException If the bucket cannot be determined from the URI.
+     */
+    private static String getBucketName(URI uri)
+    {
+        if (uri.getHost() != null) {
+            return uri.getHost();
+        }
+
+        if (uri.getUserInfo() == null) {
+            return uri.getAuthority();
+        }
+
+        throw new IllegalArgumentException("Unable to determine S3 bucket from URI.");
+    }
+
+    private static String keyFromUri(URI uri)
+    {
+        checkArgument(uri.isAbsolute(), "Uri is not absolute: %s", uri);
+        String key = nullToEmpty(uri.getPath());
+        if (key.startsWith(PATH_SEPARATOR)) {
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
+        return key;
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static AwsCredentialsProvider createAwsCredentialsProvider(ExchangeS3Config config)
+    {
+        if (config.getS3AwsAccessKey() != null && config.getS3AwsSecretKey() != null) {
+            return StaticCredentialsProvider.create(AwsBasicCredentials.create(config.getS3AwsAccessKey(), config.getS3AwsSecretKey()));
+        }
+        return DefaultCredentialsProvider.create();
+    }
+
+    private S3Client createS3Client(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3ClientBuilder clientBuilder = S3Client.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private S3AsyncClient createS3AsyncClient(AwsCredentialsProvider credentialsProvider, ClientOverrideConfiguration overrideConfig)
+    {
+        S3AsyncClientBuilder clientBuilder = S3AsyncClient.builder()
+                .credentialsProvider(credentialsProvider)
+                .overrideConfiguration(overrideConfig);
+
+        if (region != null) {
+            clientBuilder = clientBuilder.region(region);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static class S3ExchangeStorageWriter
```

Comment:
Should this interface be thread safe?

Reply:
Currently this is not thread safe, and the synchronization strategy ensures that it will only be used in a single-threaded manner.